torch/csrc/jit/codegen/cuda/lower_mem_index.cpp

```diff
@@ -0,0 +1,1222 @@
+      auto process_node = [&result, &stop_set](IterDomain* id) {
+        if (stop_set.count(id)) {
+          result.push_back(id);
+        } else {
+          auto leaves = getFlattenedSplitIds(id, stop_set);
+          result.insert(result.end(), leaves.begin(), leaves.end());
+        }
+      };
+
+      process_node(split->outer());
+      process_node(split->inner());
+      return result;
+    }
+  }
+
+  // base case where the current node is even a leaf or
+  //  not a split consumer:
+  result.push_back(split_id);
+  return result;
+}
+
+//! Returns the stride of the leading dimension given the vector of leaf
+//!  iterdomains following their order defined by their outer/inner position
+//!  as input/output of merge/split ops.
+c10::optional<int64_t> getMaybeLeadingStride(
+    std::vector<IterDomain*> leaf_ids,
+    IterDomain* leading_id = nullptr) {
+  if (leaf_ids.empty()) {
+    return c10::nullopt;
+  }
+  if (leading_id == nullptr) {
+    leading_id = leaf_ids.at(0);
+  }
+
+  bool leading_id_found = false;
+  // Utility to compute the effective stride of id and prev_id:
+  c10::optional<int64_t> result = c10::nullopt;
+  for (auto leaf_it = leaf_ids.begin(); leaf_it != leaf_ids.end(); leaf_it++) {
+    auto leaf_id = *leaf_it;
+    if (leading_id_found) {
+      if (!leaf_id->extent()->isConstInt()) {
+        // Non-constant dimensions cannot be supported.
+        return c10::nullopt;
+      }
+      result.value() *= leaf_id->extent()->evaluateInt();
+    } else {
+      if (leaf_id == leading_id) {
+        leading_id_found = true;
+        result = 1;
+      }
+    }
+  }
+  return result;
+}
+
+// Returns true if the given leaf (not checked) iterdomain,
+//  from the given tensorview's tensor domain is separable, as defined
+//  in [Note: Separability check]
+//
+// The analysis propagates backward, from the given leaf to the
+//  contig merged frontier. There are two specific patterns that
+//  this pass looks at, which are the ones that allows the index math
+//  involving the given iterdomain to be separated out and more importantly
+//  added to the rest, as memory index inlining is addition only so no need
+//  to look at other ways of hoisting and combining here:
+//
+// pattern 1: (split-only propagation)
+//       Id0 (root or contig merged id)
+//        |
+//      split
+//      /   |
+//    Id1   Id2(32)
+//            |
+//           split
+//           /   |
+//         Id3   Id4
+// In the above example {Id1, Id3, Id4} are separable.
+//
+// pattern 2: (merge-then-split propagation):
+//
+//  Id0(root)  Id1(32)
+//   |         /
+//      merge
+//       |
+//      Id2    Id3(16)
+//       |      /
+//        merge
+//          |
+//         Id4
+//          |
+//         split
+//         /  |
+//        Id5  Id6(32)
+//         |
+//       split
+//       /   |
+//     Id7   Id8(16)
+// In the case above, the leaf ids are {Id7, Id8, Id6},
+//   and Id7 is the one that's separable while the others are not.
```

Comment:
I am also confused here, but my guess is because in the index math of `Id0`, `Id1` and `Id3`, the leafs are used as:
`(i7 * 16 + i8) * 32 + i6`, where `i7` is the only number that is multiplied by both 32 and 16, so it will survive the `/32, %32, /16, %16` from the merges...